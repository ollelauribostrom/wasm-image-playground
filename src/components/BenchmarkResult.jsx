import React from 'react';
import Icon from './Icon';

function BenchmarkResult({ task, fastest, percent, avgDiff }) {
  const diff = avgDiff !== undefined ? `(${avgDiff}% difference in output pixels)` : '';
  const info = `${task}: ${fastest} was ${percent.total}% faster ${diff}`;
  return (
    <div className="benchmark-result" key={task}>
      {<Icon name="results" size="s" />}
      <span className="benchmark-result-info">{info}</span>
    </div>
  )
}

export default BenchmarkResult;
