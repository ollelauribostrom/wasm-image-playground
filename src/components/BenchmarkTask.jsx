import React from 'react';
import Icon from './Icon';
import Spinner from './Spinner';

function getStatusIcon(status) {
  if (status === 'running') {
    return (
      <div className="benchmark-task-status-icon">
        <Spinner visible={true} color="#A599FF" />
      </div>
    );
  }
  if (status === 'done') {
    return <Icon name="check" size="s" className="benchmark-task-status-icon" />;
  }
  if (status === 'error') {
    return <Icon name="cross" size="s" className="benchmark-task-status-icon" />;
  }
}

function BenchmarkTask({ id, status, info, time, average, type, iconSize = 's' }) {
  return (
    <div className="benchmark-task" key={id}>
      {<Icon name={type} size={iconSize} />}
      <span className="benchmark-info">{info}</span>
      {time ? <span className="benchmark-time">Total: {time}ms</span> : null}
      {average ? <span className="benchmark-average">Average: {average}ms</span> : null }
      {getStatusIcon(status)}
    </div>
  )
}

export default BenchmarkTask;
