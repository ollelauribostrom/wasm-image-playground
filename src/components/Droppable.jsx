import React, { Component } from 'react';

type DroppableProps = {
  onDrop: () => void,
  onDragStart?: () => void,
  onDragEnd?: () => void
};

type DroppableState = {
  dragging: boolean,
  draggingTimeout: number
};

class Droppable extends Component<DroppableProps, DroppableState> {
  droppableContainer: any;
  state = {
    dragging: false,
    draggingTimeout: null
  };
  static defaultProps = {
    onDragStart: () => {},
    onDragEnd: () => {}
  };
  componentDidMount() {
    this.droppableContainer.addEventListener('dragover', this.onDragOver);
    this.droppableContainer.addEventListener('drop', this.props.onDrop);
  }
  componentWillUnmount() {
    window.clearTimeout(this.state.draggingTimeout);
    this.droppableContainer.removeEventListener('dragover', this.onDragOver);
    this.droppableContainer.removeEventListener('drop', this.props.onDrop);
  }
  onDragOver = ev => {
    ev.preventDefault();
    window.clearTimeout(this.state.draggingTimeout);
    this.props.onDragStart();
    const draggingTimeout = window.setTimeout(() => {
      this.setState({ dragging: false });
      this.props.onDragEnd();
    }, 100);
    this.setState({ dragging: true, draggingTimeout });
  };
  render() {
    return (
      <div
        ref={ref => {
          this.droppableContainer = ref;
        }}
      >
        {this.props.children}
      </div>
    );
  }
}

export default Droppable;
