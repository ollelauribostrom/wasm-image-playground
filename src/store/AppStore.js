import createStore from 'react-waterfall';
import ColorService from '../services/ColorService';

type AppState = {|
  count: number,
  color: string
|};

const initialState: AppState = {
  count: 0,
  color: '#fff'
};

const config = {
  initialState,
  actionsCreators: {
    click: async ({ count }: AppState): Promise<AppState> => ({
      count: count + 1,
      color: await ColorService.getRandomColor()
    })
  }
};

export const { Provider, connect, actions } = createStore(config);
