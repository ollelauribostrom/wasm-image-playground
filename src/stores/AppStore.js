import createStore from 'react-waterfall';
import ImageService from '../services/ImageService';
import type { Language, Filter } from '../services/types';

type AppState = {
  image?: Image,
  history: Array<Image>,
  status?: string,
  language: Language,
  isDragging: boolean,
  isLoading: boolean,
  showBenchmarkModal: boolean
};

export const initialState: AppState = {
  image: null,
  history: [],
  status: null,
  language: 'js',
  isDragging: false,
  isLoading: false,
  showBenchmarkModal: false
};

export const config = {
  initialState,
  actionsCreators: {
    setStatus: (state: AppState, actions: any, status: boolean): AppState => {
      return { status };
    },
    setLanguage: (state: AppState, actions: any, language: Language): AppState => {
      return { language };
    },
    setLoading: (state: AppState, actions: any, isLoading: boolean): AppState => {
      return { isLoading };
    },
    setDragging: (state: AppState, actions: any, isDragging: boolean): AppState => {
      return { isDragging };
    },
    toggleBenchmarkModal: ({ showBenchmarkModal }: AppState): AppState => {
      return { showBenchmarkModal: !showBenchmarkModal };
    },
    upload: async (state: AppState, actions: any, files: FileList): AppState => {
      const image = await ImageService.upload(files);
      return { image, isLoading: false };
    },
    applyFilter: async ({ image, history }: AppState, actions: any, filter: Filter): AppState => {
      const newImage = await ImageService.applyFilter(image, filter);
      const newHistory = [...history];
      newHistory.push(image);
      return { image: newImage, history: newHistory, isLoading: false };
    },
    runBenchmark: async (): AppState => {
      const benchmarkResults = await ImageService.benchmark();
      return { benchmarkResults, isLoading: false, showBenchmarkModal: true };
    },
    undo: ({ history }: AppState): AppState => {
      const newHistory = [...history];
      const newImage = newHistory.pop();
      return { image: newImage, history: newHistory, isLoading: false };
    },
    download: ({ image }: AppState): AppState => {
      ImageService.download(image);
      return { isLoading: false };
    },
    delete: (): AppState => {
      return { image: null, history: [] };
    }
  }
};

export const { Provider, connect, actions } = createStore(config);
