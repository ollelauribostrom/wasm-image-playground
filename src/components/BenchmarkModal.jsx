import React from 'react';
import Modal from 'react-modal';
import Icon from './Icon';
import BenchmarkTask from './BenchmarkTask';
import BenchmarkResult from './BenchmarkResult';
import Label from './Label';

const modalStyle = {
  content: {
    background: "transparent",
    border: 0
  },
  overlay: {
    background: "rgba(0, 0, 0, 0.9)",
    border: 0,
    zIndex: 2,            
  }
};

function BenchmarkModal({ isOpen, title, running, results, tasks, onStart, onRestart, onClose }) {
  return (
    <Modal isOpen={isOpen} style={modalStyle} >
      <div className="benchmark-content">
        <h1 className="benchmark-heading">Benchmark: {title}</h1>
        <Icon 
          name="restore"
          size="l"
          onClick={onRestart}
          className={`benchmark-reset-icon ${!running && results.length ? '' : 'hidden'}`}
        />
        <Icon 
          name="close"
          size="l"
          onClick={onClose}
          className="benchmark-close-icon"
        />
        <Label 
          text="Run Benchmark"
          size="large"
          onClick={onStart}
          className={`benchmark-button ${!running && !results.length ? '' : 'hidden'}`} 
        />
        {tasks.map(task => <BenchmarkTask {...task} />)}
        <div className={`benchmark-results ${results.length ? '' : 'hidden'}`}>
          {results.map(result => <BenchmarkResult {...result} />)}
        </div>
      </div>
    </Modal>
  )
}

export default BenchmarkModal;