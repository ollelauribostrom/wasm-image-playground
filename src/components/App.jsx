import React from 'react';
import { Provider } from '../store/AppStore';
import ErrorBoundary from './ErrorBoundary';
import Reactor from './Reactor';
import Count from './Count';
import '../styles/main.css';

export default function App() {
  return (
    <Provider>
      <ErrorBoundary>
        <Reactor />
        <Count />
      </ErrorBoundary>
    </Provider>
  );
}
