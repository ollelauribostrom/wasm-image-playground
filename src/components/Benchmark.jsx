import React, { Component } from 'react';
import { imageConverters } from 'imutils';
import BenchmarkModal from './BenchmarkModal';
import ImageService from '../services/ImageService';
import benchmarkImages from '../../benchmark/images.json'

class Benchmark extends Component {
  state = {
    taskIndex: 0,
    tasks: [],
    taskResults: {},
    results: [],
    error: null,
    running: false,
    showResult: false
  }

  componentWillMount() {
    ImageService.on('benchmarkUpdate', this.onUpdate);
    ImageService.on('benchmarkError', this.onError);
    ImageService.on('benchmarkComplete', this.onComplete);
  }

  componentWillUnmount() {
    ImageService.removeListener('benchmarkUpdate', this.onUpdate);
    ImageService.removeListener('benchmarkError', this.onError);
    ImageService.removeListener('benchmarkComplete', this.onComplete);
  }

  restart = async () => {
    if (!this.state.running && this.state.results) {
      await this.setState({ tasks: [], results: [], error: null, taskIndex: 0, taskResults: {} });
      this.start();
    }
  }

  start = () => {
    this.props.onStart();
    this.setState({ running: true });
    this.setup();
  }

  stop = () => {
    this.props.onStop();
  }

  setup = async () => {
    try {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Loading benchmark images',
        status: 'running',
        type: 'setup'
      });
      const images = await Promise.all(benchmarkImages.files.map(async image => {
        const res = await fetch(require(`../../benchmark/${image.path}`));
        const blob = await res.blob();
        const data = await imageConverters.toUint8ClampedArray(URL.createObjectURL(blob));
        return { data, containsFace: image.containsFace }
      }));
      await this.setState({ images });
      this.addOrUpdateTask({
        id: 'SETUP',
        info: `Loaded ${images.length} benchmark images`,
        status: 'done',
      });
      this.startTask(this.props.tasks[0]);
    } catch (error) {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Error when loading benchmark images',
        status: 'error',
      });
    }
  }

  startTask = task => {
    ImageService.postMessage({
      task,
      action: this.props.benchmarkType,
      images: this.state.images,
      results: task === 'getResults' ? this.state.taskResults : null
    });
  }

  addOrUpdateTask = task => {
    const tasks = [...this.state.tasks];
    const taskIndex = tasks.findIndex(t => t.id === task.id);
    if (taskIndex > -1) {
      tasks[taskIndex] = Object.assign(tasks[taskIndex], task);
    } else {
      tasks.push(task)
    }
    this.setState({ tasks });

    if (task.status === 'done' && task.id !== 'SETUP') {
      const taskResults = Object.assign({}, this.state.taskResults);
      const currentTask = this.props.tasks[this.state.taskIndex];
      const nextTask = this.props.tasks[this.state.taskIndex + 1];
      taskResults[currentTask] = task.result;
      this.setState({ taskResults, taskIndex: this.state.taskIndex + 1 });
      if (nextTask) {
        this.startTask(nextTask);
      }
    }
  }

  onUpdate = ({ data }) => {
    this.addOrUpdateTask(data.task);
  }
  
  onError = ({ data }) => {
    this.setState({
      error: data.error,
      running: false
    });
    this.props.onStop();
  }

  onComplete = ({ data }) => {
    this.setState({ running: false, results: data.results });
    this.props.onStop();
  }

  onClose = () => {
    this.setState({
      isOpen: false
    });
  }

  render() {
    return (
      <BenchmarkModal
        isOpen={this.props.isOpen}
        running={this.state.running}
        tasks={this.state.tasks}
        results={this.state.results}
        title={this.props.title}
        onStart={this.start}
        onRestart={this.restart}
        onClose={this.props.onClose}
      />
    )
  }
}

export default Benchmark;
