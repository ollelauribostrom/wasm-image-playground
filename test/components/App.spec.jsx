import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../../src/components/App';
import { actions } from '../../src/stores/AppStore';
import { Droppable } from '../../src/components/Droppable';

jest.mock('../../src/stores/AppStore', () => ({
  connect: () => arg => arg,
  Provider: () => <div />,
  actions: {
    setLoading: jest.fn(),
    setDragging: jest.fn(),
    upload: jest.fn()
  }
}));

describe('Tests for App', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
  it('handles drop events on the Droppable component (e.target.files)', async () => {
    const wrapper = shallow(<App />);
    const onDrop = wrapper.find(Droppable).prop('onDrop');
    const mockFiles = ['fileA', 'fileB'];
    const mockEvent = { preventDefault: jest.fn(), target: { files: mockFiles } };
    await onDrop(mockEvent);
    expect(mockEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(actions.setLoading).toHaveBeenCalledTimes(1);
    expect(actions.upload).toHaveBeenCalledWith(mockFiles);
    wrapper.unmount();
  });
  it('handles drop events on the Droppable component (e.dataTransfer.files)', async () => {
    const wrapper = shallow(<App />);
    const onDrop = wrapper.find(Droppable).prop('onDrop');
    const mockFiles = ['fileA', 'fileB'];
    const mockEvent = { preventDefault: jest.fn(), target: {}, dataTransfer: { files: mockFiles } };
    actions.setLoading.mockClear();
    actions.upload.mockClear();
    await onDrop(mockEvent);
    expect(mockEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(actions.setLoading).toHaveBeenCalledTimes(1);
    expect(actions.upload).toHaveBeenCalledWith(mockFiles);
    wrapper.unmount();
  });
  it('handles dragStart events on the Droppable component', () => {
    const wrapper = shallow(<App />);
    const onDragStart = wrapper.find(Droppable).prop('onDragStart');
    onDragStart();
    expect(actions.setDragging).toHaveBeenCalledWith(true);
  });
  it('handles dragEnd events on the Droppable component', () => {
    const wrapper = shallow(<App />);
    const onDragEnd = wrapper.find(Droppable).prop('onDragEnd');
    onDragEnd();
    expect(actions.setDragging).toHaveBeenCalledWith(false);
  });
});
