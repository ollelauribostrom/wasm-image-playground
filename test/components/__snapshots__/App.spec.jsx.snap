// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for App renders correctly 1`] = `
<Provider>
  <ErrorBoundary>
    <Header />
    <Droppable
      onDragEnd={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <ImageView />
    </Droppable>
    <Footer />
  </ErrorBoundary>
</Provider>
`;
